### PR TITLE
fix visualization/draw ICP example and add warnings

### LINKED ISSUE #6887: ICP registration crashes in visualization/draw
### Checklist

- [X] I have searched for [similar issues](https://github.com/isl-org/Open3D/issues).
- [X] For Python issues, I have tested with the [latest development wheel](https://www.open3d.org/docs/latest/getting_started.html#development-version-pip).
- [X] I have checked the [release documentation](https://www.open3d.org/docs/release/) and the [latest documentation](https://www.open3d.org/docs/latest/) (for `main` branch).

### Describe the issue

Running the **`visualization/draw`** example and clicking on _ICP registration_ results in a crash about `list index out of range`.

Previously in O3D 0.14, there may have been a similar issue, see #4531.

### Steps to reproduce the bug

Run the test command as described in the documentation at https://www.open3d.org/docs/latest/getting_started.html#try-it:

```python
open3d example visualization/draw
```

Then close the windows until you are at the two pointclouds.
Then click "ICP registration (one set)" or "ICP registration (two sets)". It will crash.

### Error message
```
(open3d-dev) user@user-pc:~$ open3d example visualization/draw
***************************************************
* Open3D: A Modern Library for 3D Data Processing *
*                                                 *
* Version 0.18.0+c1b55ee                          *
* Docs    https://www.open3d.org/docs             *
* Code    https://github.com/isl-org/Open3D       *
***************************************************
Running example visualization/draw ...
FEngine (64 bits) created at 0x55e4370 (threading is enabled)
FEngine resolved backend: OpenGL
[Open3D INFO] Downloading https://github.com/isl-org/open3d_downloads/releases/download/20220201-data/BunnyMesh.ply
[Open3D INFO] Downloaded to /home/user/open3d_data/download/BunnyMesh/BunnyMesh.ply
[Open3D INFO] Downloading https://github.com/isl-org/open3d_downloads/releases/download/20220301-data/DemoICPPointClouds.zip
[Open3D INFO] Downloaded to /home/user/open3d_data/download/DemoICPPointClouds/DemoICPPointClouds.zip
[Open3D INFO] Created directory /home/user/open3d_data/extract/DemoICPPointClouds.
[Open3D INFO] Extracting /home/user/open3d_data/download/DemoICPPointClouds/DemoICPPointClouds.zip.
[Open3D INFO] Extracted to /home/user/open3d_data/extract/DemoICPPointClouds.
Traceback (most recent call last):
  File "/home/user/miniconda3/envs/open3d-dev/bin/open3d", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/user/miniconda3/envs/open3d-dev/lib/python3.11/site-packages/open3d/tools/cli.py", line 368, in main
    return args.func(subparsers.choices[args.command], args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/miniconda3/envs/open3d-dev/lib/python3.11/site-packages/open3d/tools/cli.py", line 184, in _example
    runpy.run_path(target, run_name="__main__")
  File "<frozen runpy>", line 291, in run_path
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "/home/user/miniconda3/envs/open3d-dev/lib/python3.11/site-packages/open3d/examples/visualization/draw.py", line 327, in <module>
    main()
  File "/home/user/miniconda3/envs/open3d-dev/lib/python3.11/site-packages/open3d/examples/visualization/draw.py", line 321, in main
    selections()
  File "/home/user/miniconda3/envs/open3d-dev/lib/python3.11/site-packages/open3d/examples/visualization/draw.py", line 179, in selections
    vis.draw([{
  File "/home/user/miniconda3/envs/open3d-dev/lib/python3.11/site-packages/open3d/visualization/draw.py", line 227, in draw
    gui.Application.instance.run()
  File "/home/user/miniconda3/envs/open3d-dev/lib/python3.11/site-packages/open3d/examples/visualization/draw.py", line 149, in do_icp_one_set
    source_picked = sorted(list(sets[0][source_name]),
                                ~~~~^^^
IndexError: list index out of range
```

### Expected behavior

It won't crash and perform the point cloud registration.

### Open3D, Python and System information

```markdown
- Operating system: Linux Mint 21.3 Cinnamon
- Python version: Python 3.11.9
- Open3D version: 0.18.0+c1b55ee
- System architecture: x86
- Is this a remote workstation?: no
- How did you install Open3D?: conda, latest development wheel
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by rxba**
## Type

-   [x] Bug fix (non-breaking change which fixes an issue): Fixes #6887 
-   [ ] New feature (non-breaking change which adds functionality). Resolves #
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) Resolves #

## Motivation and Context

A minor issue, but the ICP example in `examples/python/visualization/draw.py` requires the user to select point correspondences for a preliminary rough transformation before ICP registration, but users may not be aware of this. Currently, if no point correspondences are selected, the application crashes with an IndexError.

## Checklist:

-   [x] I have run `python util/check_style.py --apply` to apply Open3D **code style**
    to my code.
-   [ ] This PR changes Open3D behavior or adds new functionality.
    -   [ ] Both C++ (Doxygen) and Python (Sphinx / Google style) **documentation** is
        updated accordingly.
    -   [ ] I have added or updated C++ and / or Python **unit tests** OR included **test
        results** (e.g. screenshots or numbers) here.
-   [x] I will follow up and update the code if CI fails.
    
-   [x] For fork PRs, I have selected **Allow edits from maintainers**.

## Description
Here are the changes I've made:
* prevented the IndexError when running the ICP example without point correspondences and added a RuntimeWarning for the user to hightlight the need to select point correspondences.
* additionally prevented a KeyError when the "Source (yellow)" cloud is not contained in the first selection set and added a RuntimeWarning for this.
* refactored the existing `do_ICP_one_set` and `do_ICP_two_set` with new functions `_prep_correspondences` and `_do_ICP` to reduce duplicate code.

### CODE REVIEW

File: examples/python/visualization/draw.py [hunk 1]
```diff
@@ -142,30 +143,30 @@ def selections():
...
-
-        # Update the source geometry
-        o3dvis.remove_geometry(source_name)
-        o3dvis.add_geometry({"name": source_name, "geometry": source})
-
-    def do_icp_two_sets(o3dvis):
-        sets = o3dvis.get_selection_sets()
         source_set = sets[0][source_name]
-        target_set = sets[1][target_name]
+        if two_set:
```
Comment (ssheorey) [inline examples/python/visualization/draw.py:161]: Still get an error if the user only selects one set and asks for two_set ICP:

```py
    target_set = sets[1][target_name]
IndexError: list index out of range
```

Can you also add a check for `len(sets)==2`?

**Review by ssheorey — CHANGES_REQUESTED** [review]
Thanks for this much needed bug fix! Looks good after adding one more check (see below).

**Comment by rxba** [comment]
Thank you for the feedback! I've added your suggested check, and also found two more things that needed checking to avoid a crash, which I also added:
* add check if only one set provided to two_set ICP
* add check if no Target (blue) points are provided to single set ICP
* add check if overall number of corresponding points is mismatching and report the mismatch so that the user can add to the existing sets

**Review by ssheorey — APPROVED** [review]
Thanks @rxba ! Looks good.

**Pull request merged**